src/units/abilities.cpp

```diff
@@ -1535,7 +1535,17 @@ bool attack_type::has_weapon_ability(const std::string& special, bool special_id
 
 bool attack_type::has_special_or_ability(const std::string& special, bool special_id, bool special_tags) const
 {
-	return (has_special(special, false, special_id, special_tags) || has_weapon_ability(special, special_id, special_tags));
+	if(has_special(special, false, special_id, special_tags)){
+		return true;
+	}
+	//Now that filter_(second)attack in event supports special_id/type_active, including abilities used as weapons,
+	//these can be detected even in placeholder attacks generated to compensate for the lack of attack in defense against an attacker using a range attack not possessed by the defender.
+	//It is therefore necessary to check if the range is not empty (proof that the weapon is not a placeholder) to decide if has_weapon_ability can be returned or not.
+	bool real_weapon_ability = (range() != "") ? has_weapon_ability(special, special_id, special_tags) : false;
+	if(real_weapon_ability){
+		return true;
+	}
+	return false;
```

Comment:
No need for a `real_weapon_ability` variable. This can be written as:

```
	if(range().empty()){
		return false;
	}
	return has_weapon_ability(special, special_id, special_tags);
```

Reply:
done in https://github.com/wesnoth/wesnoth/pull/7539/commits/5126b88e5d8a4c781ccaa1f7b4ff51aaf3348738

Reply:
I'm confused - I thought that the `has_special(...)` would need to be before the `if(range().empty())`, because abilities might give extra defense even when the unit doesn't have a weapon to counterattack with.

Reply:
Fixed